internal/cloudapi/server.go

```diff
@@ -517,61 +483,47 @@ func composeStatusFromJobStatus(js *worker.JobStatus, result *worker.OSBuildJobR
 }
 
 // GetOpenapiJson handles a /openapi.json GET request
-func (server *Server) GetOpenapiJson(w http.ResponseWriter, r *http.Request) {
+func (h *apiHandlers) GetOpenapiJson(ctx echo.Context) error {
 	spec, err := GetSwagger()
 	if err != nil {
-		http.Error(w, "Could not load openapi spec", http.StatusInternalServerError)
-		return
-	}
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
-	err = json.NewEncoder(w).Encode(spec)
-	if err != nil {
-		panic("Failed to write response")
+		return echo.NewHTTPError(http.StatusInternalServerError, "Could not load openapi spec")
 	}
+	return ctx.JSON(http.StatusOK, spec)
 }
 
 // GetVersion handles a /version GET request
-func (server *Server) GetVersion(w http.ResponseWriter, r *http.Request) {
+func (h *apiHandlers) GetVersion(ctx echo.Context) error {
 	spec, err := GetSwagger()
 	if err != nil {
-		http.Error(w, "Could not load version", http.StatusInternalServerError)
-		return
+		return echo.NewHTTPError(http.StatusInternalServerError, "Could not load version")
 	}
 	version := Version{spec.Info.Version}
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
-	err = json.NewEncoder(w).Encode(version)
-	if err != nil {
-		panic("Failed to write response")
-	}
+	return ctx.JSON(http.StatusOK, version)
 }
 
 // ComposeMetadata handles a /compose/{id}/metadata GET request
-func (server *Server) ComposeMetadata(w http.ResponseWriter, r *http.Request, id string) {
+func (h *apiHandlers) ComposeMetadata(ctx echo.Context, id string) error {
 	jobId, err := uuid.Parse(id)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("Invalid format for parameter id: %s", err), http.StatusBadRequest)
-		return
+		return echo.NewHTTPError(http.StatusBadRequest, "Invalid format for parameter id: %s", err)
 	}
 
 	var result worker.OSBuildJobResult
-	status, _, err := server.workers.JobStatus(jobId, &result)
+	status, _, err := h.server.workers.JobStatus(jobId, &result)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("Job %s not found: %s", id, err), http.StatusNotFound)
-		return
+		return echo.NewHTTPError(http.StatusNotFound, "Job %s not found: %s", id, err)
 	}
 
 	var job worker.OSBuildJob
-	if _, _, _, err = server.workers.Job(jobId, &job); err != nil {
-		http.Error(w, fmt.Sprintf("Job %s not found: %s", id, err), http.StatusNotFound)
-		return
+	if _, _, _, err = h.server.workers.Job(jobId, &job); err != nil {
+		return echo.NewHTTPError(http.StatusNotFound, "Job %s not found: %s", id, err)
 	}
 
 	if status.Finished.IsZero() {
 		// job still running: empty response
-		if err := json.NewEncoder(w).Encode(new(ComposeMetadata)); err != nil {
+		if err := ctx.JSON(200, ComposeMetadata{}); err != nil {
 			panic("Failed to write response: " + err.Error())
```

Comment:
just `return ctx.JSON` here. If there's an error creating the response it'll generate a 500 error. The panic here can be removed, as it'll crash composer.